### PR TITLE
Add a convenience `define-generic` macro.

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This adds a `define-generic` macro, that targets a frequent use-case we have for `defgeneric`: one generic, one method. Using `defmethod` for these cases is quite tempting, because it's really simple to write a straigh-to-the-point
``` lisp
(defmethod name ((some specialized) args &key and more &allow-other-keys)
  "Method docs."
  (do-something))
```
instead of a noisy
``` lisp
(defgeneric name (some args &key &allow-other-keys)
  (:method ((some specialized) args &key and more &allow-other-keys)
    "Method docs."
    (do-something))
  (:documentation "Generic docs."))
```

------------------------------------------------------------------------------------------

`define-generic` relieves the `defgeneric` fatigue by optimizing the layout of things to resemble `defun`/`defmethod` and to smartly parse the arglist and options that `defgeneric` has. With it, the definition is exactly like `defmethod`
``` lisp
(define-generic name ((some specialized) args &key and more &allow-other-keys)
  "Generic docs."
  (do-something))
```
and it expands to
``` lisp
(defgeneric name (some args &key &allow-other-keys)
  (:method ((some specialized) args &key and more &allow-other-keys)
    (do-something))
  (:documentation "Generic docs."))
```

Neat, huh?

# Discussion

- I'm pretty sure that this macro works as intended, as I've tested it in lots of cases. Additional testing is welcome, though. 
- And let me know if this syntax makes sense to you!
- Does it maybe belong to `nclasses`? Any other convenience library we can contribute it to, maybe?

I'll propagate this syntax everywhere in Nyxt once we decide `define-generic` is a good shot at generics brevity.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [X] I've added the new dependencies as:
  - ASDF dependencies,
  - Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [X] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

### CODE REVIEW

**Comment by aartaka** [comment]
> I have doubts about this one. The whole point of using `defgeneric` is to specialize against several types/classes.

No, not necessarily. In our case, the main value of `defgeneric` (IIRC) is that it is better for the inspector, as it saves a definite source location for the function.

> ```lisp
> (defgeneric name (arg-list)
>   (:method ((instance class-foo)) ...)
>   (:method ((instance class-bar)) ...)
>   ...)
> ```

`define-generic` support this syntax too. I've gone a great length to make it fully (not sure, but was working in all the cases I could think of) backwards-compatible with `defgeneric` :)

More so: even this form can be enhanced with leading docstring, as in
``` lisp
(nyxt/utilities:define-generic generic-name (value &optional compact-p)
  "Some docstring"
  (:method (value &optional compact-p)
    (declare (ignore compact-p))
    (princ value))
  (:method ((value null) &optional compact-p)
    (declare (ignore compact-p))
    (princ value))
  (:method ((value string) &optional compact-p)
    (declare (ignore compact-p))
    (princ value)))
```

> There may be a single method but, even in that case, seeing it defined via `defgeneric` seems preferable.

The thing quite often is that there's only one default method. In Nyxt code, there are twenty such instances our of forty-ish `defgeneric`s.

**Comment by Ambrevar** [comment]
It's a small improvement, but it's frequent enough that it makes sense to have it.

As part of Nclasses?  I don't know.  Generic functions are some abstract MOP objects, but not exactly _classes_, so it's questionable it belongs there, but that wouldn't be too far fetched either.

Hmmm... I'd vote for yes!

File: source/utilities.lisp [hunk 1]
```diff
@@ -51,6 +51,95 @@ any expensive computation only takes place once."
...
+   (if rotate
+       (uiop/utility:strcat (princ-to-string b) a)
+       (uiop/utility:strcat a (princ-to-string b))))
+  (:method ((a integer) (b string) &key rotate)
+   (if rotate
+       (uiop/utility:strcat b (princ-to-string a))
+       (uiop/utility:strcat (princ-to-string a) b)))
+  ;; :method-combination is preserved.
+  (:method-combination progn)
+  (:documentation \"Concatenate integers to strings.\"))"
```
Comment (Ambrevar) [inline source/utilities.lisp:126]: Maybe move some of these examples to tests?  The docstring is a bit too long anyways :)

**Comment by aartaka** [comment]
> As part of Nclasses? I don't know. Generic functions are some abstract MOP objects, but not exactly _classes_, so it's questionable it belongs there, but that wouldn't be too far fetched either.

My reasoning was that `nclasses` is about syntax sugar, so `define-generic`, as an intuitive enough of a syntax sugar belongs there.

> Hmmm... I'd vote for yes!

You mean moving it to `nclasses`? I'll open a PR there then.

File: source/utilities.lisp (hunk 1 above)
Comment (aartaka) [reply source/utilities.lisp:126]: Two reasons to retain these as they are:
- `define-generic` has quite a set of behaviors, so it's important to showcase them (but I guess that having one long example is better than two noisy ones).
- I feel like testing macroexpansions may be dangerous and implementation-dependent. Haven't tried it, though.

**Comment by Ambrevar** [comment]
> > 
> 
> You mean moving it to `nclasses`? I'll open a PR there then.

Yes, thanks.

**Comment by aartaka** [comment]
Closing in favor of https://github.com/atlas-engineer/nclasses/pull/11

**Pull request closed without merging**